plugins/woocommerce/includes/admin/class-wc-admin-menus.php

```diff
@@ -246,29 +245,66 @@ public function menu_highlight() {
 				break;
 		}
+
+		// Highlight "Add new order" submenu when on the new order page.
+		$page   = isset( $_GET['page'] ) ? sanitize_text_field( wp_unslash( $_GET['page'] ) ) : ''; // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		$action = isset( $_GET['action'] ) ? sanitize_text_field( wp_unslash( $_GET['action'] ) ) : ''; // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+		if ( 'wc-orders' === $page && 'new' === $action ) {
+			$submenu_file = \Automattic\WooCommerce\Utilities\OrderUtil::get_order_admin_new_url(); // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+		}
+	}
+
+	/**
+	 * Generate order count badge HTML.
+	 *
+	 * @param int $order_count The number of processing orders.
+	 * @return string The badge HTML.
+	 */
+	private function get_order_count_badge( $order_count ) {
+		return ' <span class="awaiting-mod update-plugins count-' . esc_attr( $order_count ) . '"><span class="processing-count">' . number_format_i18n( $order_count ) . '</span></span>';
 	}
 
 	/**
 	 * Adds the order processing count to the menu.
 	 */
 	public function menu_order_count() {
-		global $submenu;
+		global $submenu, $menu;
 
+		// Always remove 'WooCommerce' sub menu item.
 		if ( isset( $submenu['woocommerce'] ) ) {
-			// Remove 'WooCommerce' sub menu item.
 			unset( $submenu['woocommerce'][0] );
+		}
 
-			// Add count if user has access.
-			if ( apply_filters( 'woocommerce_include_processing_order_count_in_menu', true ) && current_user_can( 'edit_others_shop_orders' ) ) {
-				$order_count = apply_filters( 'woocommerce_menu_order_count', wc_processing_order_count() );
-
-				if ( $order_count ) {
-					foreach ( $submenu['woocommerce'] as $key => $menu_item ) {
-						if ( 0 === strpos( $menu_item[0], _x( 'Orders', 'Admin menu name', 'woocommerce' ) ) ) {
-							$submenu['woocommerce'][ $key ][0] .= ' <span class="awaiting-mod update-plugins count-' . esc_attr( $order_count ) . '"><span class="processing-count">' . number_format_i18n( $order_count ) . '</span></span>'; // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
-							break;
-						}
-					}
+		/**
+		 * Filters whether to include the processing order count in the menu.
+		 *
+		 * @since 2.5.0
+		 *
+		 * @param bool $include_count Whether to include the count. Default true.
+		 */
+		if ( ! apply_filters( 'woocommerce_include_processing_order_count_in_menu', true ) || ! current_user_can( 'edit_others_shop_orders' ) ) {
+			return;
+		}
+
+		/**
+		 * Filters the order count displayed in the menu.
+		 *
+		 * @since 3.9.0
+		 *
+		 * @param int $order_count The order count.
+		 */
+		$order_count = apply_filters( 'woocommerce_menu_order_count', wc_processing_order_count() );
+
+		if ( $order_count ) {
+			// Determine the menu slug to search for based on HPOS status.
+			$orders_menu_slug = \Automattic\WooCommerce\Utilities\OrderUtil::custom_orders_table_usage_is_enabled() ? 'wc-orders' : 'edit.php?post_type=shop_order';
```

Comment:
Not a huge problem but I wonder if we could sort of 'soft deprecate' this code and keep it only for the case when HPOS is disabled. When HPOS is enabled we fully control the registration of the menu item and can add the order count as necessary, avoiding the unnecessary iteration across `$menu` items that happens below.